### PR TITLE
Upgrade to Rust 1.82 toolchain

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary
Upgrade the developer toolchain. This does not bump the minimal MSRV.

## Test Plan

`cargo test`

### CODE REVIEW

**Review by AlexWaygood — APPROVED** [review]
LGTM, but you have a merge conflict which is preventing CI from running :-)

**Review by dhruvmanila — APPROVED** [review]

File: crates/red_knot_server/src/server.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 //! Scheduling, I/O, and API endpoints.
 
 use std::num::NonZeroUsize;
+#[allow(deprecated)]
```
Comment (AlexWaygood) [inline crates/red_knot_server/src/server.rs:4]:
```suggestion
#[allow(deprecated)]  // Our MSRV doesn't allow us to use the new name
```

File: crates/red_knot_server/src/server.rs [hunk 2]
```diff
@@ -119,6 +120,7 @@ impl Server {
     }
 
     pub(crate) fn run(self) -> crate::Result<()> {
+        #[allow(deprecated)]
```
Comment (AlexWaygood) [inline crates/red_knot_server/src/server.rs:123]:
```suggestion
        #[allow(deprecated)]  // Our MSRV doesn't allow us to use the new name
```

File: crates/ruff_server/src/server.rs [hunk 3]
```diff
@@ -6,6 +6,7 @@ use lsp_types::InitializeParams;
 use lsp_types::WorkspaceFolder;
 use std::num::NonZeroUsize;
 use std::ops::Deref;
+#[allow(deprecated)]
```
Comment (AlexWaygood) [inline crates/ruff_server/src/server.rs:9]:
```suggestion
#[allow(deprecated)]  // Our MSRV doesn't allow us to use the new name
```

File: crates/ruff_server/src/server.rs [hunk 4]
```diff
@@ -125,6 +126,7 @@ impl Server {
     }
 
     pub fn run(self) -> crate::Result<()> {
+        #[allow(deprecated)]
```
Comment (AlexWaygood) [inline crates/ruff_server/src/server.rs:129]:
```suggestion
        #[allow(deprecated)]  // Our MSRV doesn't allow us to use the new name
```

**Review by AlexWaygood — COMMENTED** [review]
I'd maybe leave some comments explaining why we're ignoring the deprecation warnings:

**Comment by AlexWaygood** [comment]
> Merging #13808 will **degrade performances by 14.01%**

that doesn't look good...

**Comment by MichaReiser** [comment]
Hmm nope

**Pull request merged**